File: backend/src/types/interview.py
from enum import Enum
from typing import Optional
from datetime import datetime
from pydantic import BaseModel, Field


class InterviewState(str, Enum):
    IDLE = "idle"
    STARTED = "started"
    QUESTIONING = "questioning"
    EVALUATING = "evaluating"
    COMPLETE = "complete"


class ExperienceLevel(str, Enum):
    JUNIOR = "junior"
    MID = "mid"
    SENIOR = "senior"
    STAFF = "staff"


class QuestionType(str, Enum):
    CONCEPTUAL = "conceptual"
    CODING = "coding"
    BEHAVIORAL = "behavioral"
    SCENARIO = "scenario"


class Question(BaseModel):
    id: str
    topic: str
    difficulty: str
    question_type: QuestionType
    experience_level: str
    question_text: str
    follow_up_texts: list[str] = Field(default_factory=list)
    rubric: dict = Field(default_factory=dict)
    tags: list[str] = Field(default_factory=list)


class TurnRecord(BaseModel):
    turn_idx: int
    speaker: str
    text: str
    timestamp: str
    question_id: Optional[str] = None


class ScoreUpdate(BaseModel):
    topic: str
    score: float = Field(ge=0, le=10)
    reasoning: str


class QuestionResult(BaseModel):
    question_id: str
    question_text: str
    topic: str
    answer_text: str
    score: Optional[float] = None
    score_reasoning: Optional[str] = None
    follow_up_count: int = 0
    time_spent_seconds: Optional[int] = None


class Evaluation(BaseModel):
    overall_score: float = Field(ge=0, le=10)
    recommendation: str
    strengths: list[str] = Field(default_factory=list)
    weaknesses: list[str] = Field(default_factory=list)
    summary: str
    per_question: list[QuestionResult] = Field(default_factory=list)
    topic_scores: dict[str, float] = Field(default_factory=dict)


class SessionState(BaseModel):
    session_id: str
    state: InterviewState = InterviewState.IDLE
    candidate_name: str
    job_role: str
    experience_level: ExperienceLevel
    required_skills: list[str] = Field(default_factory=list)
    questions: list[Question] = Field(default_factory=list)
    current_question_idx: int = 0
    transcript: list[TurnRecord] = Field(default_factory=list)
    question_results: list[QuestionResult] = Field(default_factory=list)
    running_scores: dict[str, float] = Field(default_factory=dict)
    flags: list[str] = Field(default_factory=list)
    follow_up_count: int = 0
    started_at: Optional[str] = None
    ended_at: Optional[str] = None
    evaluation: Optional[Evaluation] = None


class FinalReport(BaseModel):
    session_id: str
    candidate_name: str
    job_role: str
    experience_level: str
    evaluation: Evaluation
    transcript: list[TurnRecord]
    started_at: Optional[str]
    ended_at: Optional[str]
    duration_seconds: Optional[int]

File: frontend/src/types/interview.ts
export type InterviewState =
  | "idle"
  | "started"
  | "questioning"
  | "evaluating"
  | "complete";

export type ExperienceLevel = "junior" | "mid" | "senior" | "staff";

export interface StartInterviewRequest {
  candidate_name: string;
  job_role: string;
  experience_level: ExperienceLevel;
  required_skills: string[];
}

export interface StartInterviewResponse {
  session_id: string;
  state: InterviewState;
  question_text: string;
  question_number: number;
  total_questions: number;
  topic: string;
  candidate_name: string;
}

export interface SubmitAnswerRequest {
  session_id: string;
  answer: string;
}

export interface SubmitAnswerResponse {
  session_id: string;
  state: InterviewState;
  score: number | null;
  score_reasoning: string | null;
  next_question: string | null;
  question_number: number | null;
  total_questions: number | null;
  topic: string | null;
  is_complete: boolean;
  feedback: string | null;
}

export interface QuestionResult {
  question_id: string;
  question_text: string;
  topic: string;
  answer_text: string;
  score: number | null;
  score_reasoning: string | null;
  follow_up_count: number;
}

export interface TranscriptTurn {
  turn_idx: number;
  speaker: string;
  text: string;
  timestamp: string;
  question_id: string | null;
}

export interface ReportResponse {
  session_id: string;
  candidate_name: string;
  job_role: string;
  experience_level: string;
  overall_score: number;
  recommendation: string;
  strengths: string[];
  weaknesses: string[];
  summary: string;
  per_question: QuestionResult[];
  topic_scores: Record<string, number>;
  transcript: TranscriptTurn[];
  started_at: string | null;
  ended_at: string | null;
  duration_seconds: number | null;
}

export interface ApiError {
  error?: string;
  detail?: string | { msg: string; type: string }[];
}
